src/utils/graphql/common.ts

```diff
@@ -0,0 +1,124 @@
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+import Arweave from 'arweave';
+import { GQLNodeInterface } from 'warp-contracts';
+
+import { EvaluationTimeoutError } from '../../common/error.js';
+import { DataProtocolTransaction } from '../../types.js';
+
+export const MAX_REQUEST_SIZE = 100;
+
+/**
+ * @param arweave - arweave instance to perform gql request
+ * @param address - address of the wallet to get transactions for
+ * @param protocolTags - the data protocol to filter by
+ * @param signal - signal to abort the request
+ * @returns @type {Promise<DataProtocolTransaction[]>} - returns the list of data protocol txs, for e.g smartweave contracts
+ */
+export async function getSmartweaveContractsFromGQL({
+  address,
+  arweave,
+  signal,
+}: {
+  arweave: Arweave;
+  address: string;
+  signal?: AbortSignal;
+}): Promise<DataProtocolTransaction[]> {
+  let hasNextPage = false;
+  let cursor: string | undefined;
+  const protocolTxs = new Set<DataProtocolTransaction>();
+  do {
+    const queryObject = {
+      query: `
+      { 
+          transactions (
+              owners:["${address}"]
+              tags: [
+                      {
+                        name: 'App-Name',
+                        values: ['SmartWeaveContract'],
+                      },
+                    ],
+              sort: HEIGHT_DESC,
+              first: ${MAX_REQUEST_SIZE},
+              bundledIn: null,
+              ${cursor !== undefined ? `after: "${cursor}"` : ''}
+          ) {
+              pageInfo {
+                  hasNextPage
+              }
+              edges {
+                  cursor
+                  node {
+                      id
+                      block {
+                          height
+                      }
+                      tags {
+                          name
+                          value
+                      }
+                  }
+              }
+          }
+      }`,
+    };
+
+    if (signal?.aborted) {
+      throw new EvaluationTimeoutError('GraphQL request was aborted.');
+    }
+
+    const { status, ...response } = await arweave.api.post(
+      '/graphql',
+      queryObject,
+      {
+        signal,
+      },
+    );
+
+    if (status == 429) {
+      // Rate limited, wait for 10 seconds and retry
+      await new Promise((resolve) => setTimeout(resolve, 10_000));
+      continue;
+    }
```

Comment:
it would be nice if you could provide a retry strategy via the params, not necessary for now, but can add as a TODO

Reply:
sure - will add the todo. In the future what I would like to see is a specific gql provider that consumes queryBuilder() and has that retry stuff internally.